sql: remove the artifact of canceling the txn-scoped context

yuzefovich opened the pull request:
This commit removes an old artifact of having a txn-scoped context
cancellation that is performed when finishing the txn. As Andrei points
out, this txn-scoped cancellation is likely a leftover from ancient
times and is no longer needed. In particular, this also fixes the bug of
using the span after it was finished (which would occur with high
vmodule on `context.go` file).

Fixes: #83739.

Release note: None

mgartner reviewed (approved):
Nice find! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84628#-:-N7ItT0ADKNO6DzTAW3o:bli3xsu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84628)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/exec_util.go` line 2294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84628#-N7IwsMH5C6PDhg8TrJ1:-N7IwsMH5C6PDhg8TrJ2:b-qjrdrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/39b7178450a73f7e64d98816a30228974894b04e/pkg/sql/exec_util.go#L2294)):*
> ```Go
> 		// context cancellation). To go around that, we pass the responsibility
> 		// of finishing the span to the cancellation function.
> 		cancelTxnCtx := st.ex.state.cancel
> ```

I think finishing `sp` here was better. I think a better fix would be to call `st.ex.state.cancel` here, and then overwrite it.
But, actually, I don't understand why `ex.state.cancel` exists - why do we have a transaction-level cancelation? There's logically no such thing as canceling a transaction, is there? I think the only use of `state.cancel` is `CANCEL QUERY`, which could do with a statement-level cancel func. A statement-level func I think could be scoped to `execStmtInOpenState`. Then we wouldn't need this more troublesome and unclearly scoped transaction-level cancelation.

The transaction-level cancelation might have been a left-over from an ancient time when
a) we had parallel statements (`returning nothing`)
b) the TxnCoordSender's heartbeat loop was using a transaction-scoped SQL ctx.
Neither of these are true any more.

WDYT?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: remove the artifact of canceling the txn-scoped context

This commit removes an old artifact of having a txn-scoped context
cancellation that is performed when finishing the txn. As Andrei points
out, this txn-scoped cancellation is likely a leftover from ancient
times and is no longer needed. In particular, this also fixes the bug of
using the span after it was finished (which would occur with high
vmodule on `context.go` file).

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84628)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @mgartner)

___
*[`pkg/sql/exec_util.go` line 2294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84628#-N7IwsMH5C6PDhg8TrJ1:-N7J78TZ1eaGO_OCRk76:bkd88bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/39b7178450a73f7e64d98816a30228974894b04e/pkg/sql/exec_util.go#L2294)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think finishing `sp` here was better. I think a better fix would be to call `st.ex.state.cancel` here, and then overwrite it.
But, actually, I don't understand why `ex.state.cancel` exists - why do we have a transaction-level cancelation? There's logically no such thing as canceling a transaction, is there? I think the only use of `state.cancel` is `CANCEL QUERY`, which could do with a statement-level cancel func. A statement-level func I think could be scoped to `execStmtInOpenState`. Then we wouldn't need this more troublesome and unclearly scoped transaction-level cancelation.

The transaction-level cancelation might have been a left-over from an ancient time when
a) we had parallel statements (`returning nothing`)
b) the TxnCoordSender's heartbeat loop was using a transaction-scoped SQL ctx.
Neither of these are true any more.

WDYT?
</blockquote></details>

I think you are right on all points, so I implemented your suggestion. Let's see what the CI says, but I'm optimistic that nothing will break :crossed_fingers: 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Looks like CI is happy, and I like this approach more. PTAL.

andreimatei reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84628#-:-N7ReXNu1RnT1VeabNqA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @mgartner)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
